providers/flagd/pkg/service/http/service.go

```diff
@@ -0,0 +1,135 @@
+	resMess := schemaV1.ResolveStringResponse{}
+	err := s.FetchFlag(url, context, &resMess)
+	if err != nil {
+		return &schemaV1.ResolveStringResponse{
+			Reason: models.ErrorReason,
+		}, err
+	}
+	return &resMess, nil
+}
+
+// ResolveNumber handles the flag evaluation response from the flags/{flagKey}/resolve/number endpoint
+func (s *HTTPService) ResolveNumber(flagKey string, context of.EvaluationContext) (*schemaV1.ResolveNumberResponse, error) {
+	url := fmt.Sprintf("%s://%s:%d/flags/%s/resolve/number", s.HTTPServiceConfiguration.Protocol, s.HTTPServiceConfiguration.Host, s.HTTPServiceConfiguration.Port, flagKey)
+	resMess := schemaV1.ResolveNumberResponse{}
+	err := s.FetchFlag(url, context, &resMess)
+	if err != nil {
+		return &schemaV1.ResolveNumberResponse{
+			Reason: models.ErrorReason,
+		}, err
+	}
+	return &resMess, nil
+}
+
+// ResolveObject handles the flag evaluation response from the flags/{flagKey}/resolve/object endpoint
+func (s *HTTPService) ResolveObject(flagKey string, context of.EvaluationContext) (*schemaV1.ResolveObjectResponse, error) {
+	url := fmt.Sprintf("%s://%s:%d/flags/%s/resolve/object", s.HTTPServiceConfiguration.Protocol, s.HTTPServiceConfiguration.Host, s.HTTPServiceConfiguration.Port, flagKey)
+	resMess := schemaV1.ResolveObjectResponse{}
+	err := s.FetchFlag(url, context, &resMess)
+	if err != nil {
+		return &schemaV1.ResolveObjectResponse{
+			Reason: models.ErrorReason,
+		}, err
+	}
+	return &resMess, nil
+}
+
+// FlagKey handles the calling and parsing of the flag evaluation endpoints.
+// Argument p should be a pointer to a valid Resolve{type}Response struct for unmarshalling the response, e.g ResolveObjectResponse{}.
+func (s *HTTPService) FetchFlag(url string, ctx of.EvaluationContext, p interface{}) error {
+	body, err := json.Marshal(flattenContext(ctx))
+	if err != nil {
+		log.Error(err)
+		return errors.New(models.ParseErrorCode)
+	}
+	res, err := s.Client.Request("POST", url, bytes.NewBuffer(body))
+	if err != nil {
+		log.Error(err)
+		return errors.New(models.GeneralErrorCode)
+	}
+
+	b, err := io.ReadAll(res.Body)
+	if err != nil {
+		log.Error(err)
+		return errors.New(models.GeneralErrorCode)
+	}
+	err = json.Unmarshal(b, p)
+	if err != nil {
+		log.Error(err)
+		if res.StatusCode == 200 {
+			return errors.New(models.ParseErrorCode)
+		}
+	}
+	if res.StatusCode == 200 {
+		return nil
+	}
+	if res.StatusCode == 500 {
+		return errors.New(models.GeneralErrorCode)
+	}
+
+	errRes := schemaV1.ErrorResponse{}
+	err = json.Unmarshal(b, &errRes)
+	if err != nil {
+		log.Error(err)
+		return errors.New(models.ParseErrorCode)
+	}
+
+	if errRes.ErrorCode != "" {
+		return errors.New(errRes.ErrorCode)
+	}
+
+	return errors.New(models.GeneralErrorCode)
+}
+
+func flattenContext(ctx of.EvaluationContext) map[string]interface{} {
+	if ctx.TargetingKey != "" {
+		ctx.Attributes["TargettingKey"] = ctx.TargetingKey
+	}
+	if ctx.Attributes == nil {
+		return map[string]interface{}{}
+	}
```

Comment:
if ctx.Attributes is indeed `nil` the call at line `129` would produce a panic.
You could just exchange the order of the two if statement (checking for nil of ctx.Attributes frst)

Reply:
You're right, the reason I wrote it out in this way was in the case of the `EvaluationContext` definition growing and the `if ctx.Attributes == nil` needing to be evaluated multiple times, however at the moment there isn't any need for this and can refactor back when/if the definition is expanded

Reply:
Sorry, I think i created a bit of confusion, what I meant is something like this:
```
func flattenContext(ctx of.EvaluationContext) map[string]interface{} {
	if ctx.Attributes == nil {
		return map[string]interface{}{}
	}

	if ctx.TargetingKey != "" {
		ctx.Attributes["TargetingKey"] = ctx.TargetingKey
	}

	return ctx.Attributes
}
```
it's a bit of a nitpick but I think that's the cleanest logic:  if `ctx.Attributes` is `nil`, it makes no sense to search for a specific key in it. So you can just return early from this func :)

Reply:
oh sorry, the purpose of this function is to return a single `map[string]interface{}` from the `EvaluationContext` struct. The struct definition defines `TargetingKey` in order to ensure that it remains constant across vendors. In the instance that `ctx.Attributes` is nil but the `TargetingKey` is not, then the returned `map[string]interface{}` must still contain this field.
```
EvaluationContext{
    TargetingKey: "James",
    Attributes: nil,
}
```
should give
```
map[string]interface{}{
    "TargetingKey": "james",
}
```
Its a bit of an odd situation and I've raised an issue regarding it [here](https://github.com/open-feature/golang-sdk/issues/30#issue-1322347254)

Reply:
ah understood, now also the func name makes more sense to me :D  thx :)